core/src/test/java/org/apache/iceberg/TestManifestReaderStats.java

```diff
@@ -0,0 +1,172 @@
+import java.util.Map;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Types;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestManifestReaderStats extends TableTestBase {
+  @Parameterized.Parameters(name = "formatVersion = {0}")
+  public static Object[] parameters() {
+    return new Object[] { 1, 2 };
+  }
+
+  public TestManifestReaderStats(int formatVersion) {
+    super(formatVersion);
+  }
+
+  private static final Map<Integer, Long> VALUE_COUNT = ImmutableMap.of(3, 3L);
+  private static final Map<Integer, Long> NULL_VALUE_COUNTS = ImmutableMap.of(3, 0L);
+  private static final Map<Integer, Long> NAN_VALUE_COUNTS = ImmutableMap.of(3, 1L);
+  private static final Map<Integer, ByteBuffer> LOWER_BOUNDS =
+      ImmutableMap.of(3, Conversions.toByteBuffer(Types.IntegerType.get(), 2));
+  private static final Map<Integer, ByteBuffer> UPPER_BOUNDS =
+      ImmutableMap.of(3, Conversions.toByteBuffer(Types.IntegerType.get(), 4));
+
+  private static final Metrics METRICS = new Metrics(3L, null,
+      VALUE_COUNT, NULL_VALUE_COUNTS, NAN_VALUE_COUNTS, LOWER_BOUNDS, UPPER_BOUNDS);
+
+  private static final DataFile FILE = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-a.parquet")
+      .withFileSizeInBytes(10)
+      .withPartitionPath("data_bucket=0") // easy way to set partition data for now
+      .withRecordCount(3)
+      .withMetrics(METRICS)
+      .build();
+
+  @Test
+  public void testReadIncludesFullStats() throws IOException {
+    ManifestFile manifest = writeManifest(1000L, FILE);
+    try (ManifestReader<DataFile> reader = ManifestFiles.read(manifest, FILE_IO)) {
+      CloseableIterable<ManifestEntry<DataFile>> entries = reader.entries();
+      ManifestEntry<DataFile> entry = entries.iterator().next();
+      assertFullStats(entry.file());
+    }
+  }
+
+  @Test
+  public void testReadWithFilterIncludesFullStats() throws IOException {
+    ManifestFile manifest = writeManifest(1000L, FILE);
+    try (ManifestReader<DataFile> reader = ManifestFiles.read(manifest, FILE_IO)
+        .filterRows(Expressions.equal("id", 3))) {
+      CloseableIterable<ManifestEntry<DataFile>> entries = reader.entries();
+      ManifestEntry<DataFile> entry = entries.iterator().next();
+      assertFullStats(entry.file());
+    }
+  }
+
+  @Test
+  public void testReadEntriesWithFilterAndSelectIncludesFullStats() throws IOException {
+    ManifestFile manifest = writeManifest(1000L, FILE);
+    try (ManifestReader<DataFile> reader = ManifestFiles.read(manifest, FILE_IO)
+        .select(ImmutableSet.of("record_count"))
```

Comment:
If I change this `record_count` to something else it will result in NPE due to [`InclusiveMetrisEvaluator.eval`](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java#L83)  needing record count, however `STATS_COLUMNS` in manifest reader [doesn't have it](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/ManifestReader.java#L55-L56). I know the reader normally will only be used internally so we don't expect to run into this often, but wonder if we want to ensure `record_count` is always added when [populating stats](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/ManifestReader.java#L292).

Reply:
Yes, I think we do. Maybe we should do that in a separate update, though.

Reply:
Sounds good, I'll create a separate pr for that

Reply:
PR: https://github.com/apache/iceberg/pull/1820